### PR TITLE
Add support for rendering uninterpreted/"raw" messages

### LINKED ISSUE #8: Ability to write raw Messages
Thanks a lot for writing `prodash`! Over in https://github.com/pantsbuild/pants, we have log formatting and construction that we'd like to stick with while considering moving to `prodash::line` for rendering running work.

To do that without tearing down the UI for each "raw" stderr output, it seems like it might makes sense to support sending "raw" (unfiltered/unleveled) `Messages`, which would not be interpreted by the renderer at all, except to dump them.

Would a patch to support un-interpreted `Messages` be welcome?

### PR DESCRIPTION
**Pull request by stuhood**
Fixes #8.

### CODE REVIEW

File: src/render/line/draw.rs [hunk 1]
```diff
@@ -52,42 +52,52 @@ fn messages(
...
-        tokens.push(brush.style(Style::default().dimmed()).paint(format!(
-            "{:>fill_size$}{}",
-            "",
-            origin,
-            fill_size = current_maximum - message_block_len,
-        )));
-        tokens.push(" ".into());
-        tokens.push(brush.style(color.bold()).paint(message));
+
         let message_block_count = block_count_sans_ansi_codes(&tokens);
```
Comment (stuhood) [inline src/render/line/draw.rs:101]: Hm: it looks like `unicode-width` does not ignore ANSI control characters, so raw messages containing ANSI characters do not fully overdraw. See https://github.com/unicode-rs/unicode-width/issues/24.

File: src/render/line/draw.rs [hunk 2]
```diff
@@ -204,8 +214,35 @@ fn newline_with_overdraw(
...
+
+        fn execute(&mut self, byte: u8) {
+            if byte == b'\n' {
+                self.0 += 1;
+            }
+        }
+    }
+
+    let mut block_counter = BlockCounter(0);
+    let mut parser = vte::Parser::new();
```
Comment (stuhood) [inline src/render/line/draw.rs:239]: The `Parser` is fairly hefty, and should likely be lifted onto the `State` for reuse across messages.

**Comment by Byron** [comment]
Thanks a lot, all this looks great!

Since VTE is needed for raw messages to calculate their size correctly, I am wondering if there could be a feature-toggle for the `Raw` part of the envelope. Without it, it's essentially all like it was before without another dependency.

Is this feasible?

**Comment by stuhood** [comment]
> Since VTE is needed for raw messages to calculate their size correctly, I am wondering if there could be a feature-toggle for the `Raw` part of the envelope. Without it, it's essentially all like it was before without another dependency.
> 
> Is this feasible?

_Technically_ `vte` is required for any `Message`, since we don't validate that the message strings don't contain ANSI escapes... but I think that it is probably fine to just document that `Item::message` `Strings` must not contain any.

But yes: I agree that it would be good to avoid introducing the overhead of the `vte` parse wherever possible: can add a feature, and will probably also see about making the `vte` parse conditional on the message type.

**Comment by Byron** [comment]
> Technically vte is required for any Message, since we don't validate that the message strings don't contain ANSI escapes... but I think that it is probably fine to just document that Item::message Strings must not contain any.

It's indeed possible to feature-toggle the Raw variant of the `Envelope` which would allow everything related to it to only become available if the, say, `raw-message` feature is enabled. This would then pull in the `vte` crate. Interestingly, this would then make the `vte` feature toggle available as well which would allow it to be used for handling non-raw messages as well.

Before doing any of this, I hope it's possible to make `make tests` pass, so that the new feature toggle can be tested as well using that machinery.

**Comment by Byron** [comment]
Closing due to inactivity.

**Pull request closed without merging**